crates/valence/examples/chat.rs

```diff
@@ -0,0 +1,116 @@
+use bevy_app::App;
+use tracing::warn;
+use valence::client::despawn_disconnected_clients;
+use valence::client::event::{default_event_handler, ChatCommand, ChatMessage};
+use valence::prelude::*;
+
+const SPAWN_Y: i32 = 64;
+const ARENA_RADIUS: i32 = 32;
+
+pub fn main() {
+    tracing_subscriber::fmt().init();
+
+    App::new()
+        .add_plugin(ServerPlugin::new(()))
+        .add_startup_system(setup)
+        .add_system_to_stage(EventLoop, default_event_handler)
+        .add_system_to_stage(EventLoop, handle_message_events)
+        .add_system_to_stage(EventLoop, handle_command_events)
+        .add_system(init_clients)
+        .add_system(despawn_disconnected_clients)
+        .add_system_set(PlayerList::default_system_set())
+        .add_system(teleport_oob_clients)
+        .run();
+}
+
+fn setup(world: &mut World) {
+    let mut instance = world
+        .resource::<Server>()
+        .new_instance(DimensionId::default());
+
+    for z in -5..5 {
+        for x in -5..5 {
+            instance.insert_chunk([x, z], Chunk::default());
+        }
+    }
+
+    // Create circular arena.
+    for z in -ARENA_RADIUS..ARENA_RADIUS {
+        for x in -ARENA_RADIUS..ARENA_RADIUS {
+            let dist = f64::hypot(x as _, z as _) / ARENA_RADIUS as f64;
+
+            if dist > 1.0 {
+                continue;
+            }
+
+            let block = if rand::random::<f64>() < dist {
+                BlockState::STONE
+            } else {
+                BlockState::DEEPSLATE
+            };
+
+            for y in 0..SPAWN_Y {
+                instance.set_block_state([x, y, z], block);
+            }
+        }
+    }
+
+    world.spawn(instance);
+}
+
+fn init_clients(
+    mut clients: Query<&mut Client, Added<Client>>,
+    instances: Query<Entity, With<Instance>>,
+) {
+    for mut client in &mut clients {
+        client.set_position([0.0, SPAWN_Y as f64, 0.0]);
+        client.set_instance(instances.single());
+        client.set_game_mode(GameMode::Adventure);
+        client.send_message("Welcome to Valence! Talk about something.".italic());
+    }
+}
+
+fn handle_message_events(mut clients: Query<&mut Client>, mut messages: EventReader<ChatMessage>) {
+    for message in messages.iter() {
+        let Ok(client) = clients.get_component::<Client>(message.client) else {
+            warn!("Unable to find client for message: {:?}", message);
+            continue;
+        };
+
+        let message = message.message.to_string();
+
+        let formatted = format!("<{}>: ", client.username())
+            .bold()
+            .color(Color::YELLOW)
+            + message.into_text().not_bold().color(Color::WHITE);
+
+        // TODO! Is there a more efficient way to do this?
```

Comment:
Talking about efficiency, there is a way to for each the clients in a parallel way:

```rust
clients.par_for_each_mut(16, |mut client| {
    client.send_message(formatted.clone());
})
```

But there is a doubt about what the batch size should be used. A similar approach: https://github.com/valence-rs/valence/blob/50018a52bf31eb5f1cacf8ede7c041e2eb83eafd/crates/valence/src/client.rs#L630-L631

Reply:
Technically, the most efficient way to do this is for valence to have a broadcast method that avoids re-serializing the packet for every client it needs to send it to. But this is perfectly fine for an example.

Reply:
I mean, if we were going for the most efficient possible, we could do it outside the event loop to avoid waiting for the next tick. That might even be useful for things that interact with an external database, running it outside of the loop to avoid blocking.

I'll update this with `par_for_each_mut` in the morning. 🙂

Reply:
> what the batch size should be used

I think this probably depends mainly on the amount of CPU cores and the amount of items, but it might also depend on eg. CPU cache size. There's probably no perfect number :woman_shrugging: